File: content_tool/observability/event_log.py
"""Verbose, persisted per-step event log.

Every SSE event that flows through a run/batch emit choke point is mirrored
into ``content_tool.run_event_logs`` so a finished or failed stream can be
debugged after the fact (the live SSE history is in-memory only and lost on
restart).

This module owns two things:

1. The PURE derivation helpers (``derive_step``, ``derive_level``,
   ``cap_payload``, ``derive_duration_ms``) that turn an SSE envelope into a
   table row. They are unit-tested directly and MUST stay byte-compatible with
   the Workers-native port (the frontend reads the same row shape from both).
2. ``RunEventLogWriter`` — an async, batched, fire-and-forget writer. It is a
   pure side-channel: a DB failure is logged and the row dropped, never raised
   into the emit path, so persistence can never break streaming.
"""

from __future__ import annotations

import asyncio
import json
import os
from collections.abc import Awaitable, Callable
from contextvars import ContextVar
from datetime import UTC, datetime
from typing import Any, cast
from uuid import uuid4

import structlog
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

logger = structlog.get_logger(__name__)

# Event-string verbs whose trailing segment is a lifecycle marker, not a step
# name. ``segments[:-1][-1]`` then yields the step.
_VERBS: frozenset[str] = frozenset(
    {"start", "done", "error", "thinking", "completed", "interrupted"}
)

# Payload size guards (bytes). Mirror the Workers port exactly.
MAX_BYTES = 16384
MAX_FIELD = 2048

# Batched-write tuning.
_FLUSH_THRESHOLD = 50
_FLUSH_INTERVAL_S = 0.5

SessionFactory = Callable[[], Any]


def derive_step(event: str) -> str | None:
    """Derive the ``step`` column from an event string per the shared contract."""
    segments = event.split(".")
    if segments and segments[-1] in _VERBS:
        return segments[:-1][-1] if len(segments) > 1 else None
    return segments[-1] if segments else None


def derive_level(event: str) -> str:
    """Derive the ``level`` column from an event string per the shared contract."""
    if event.endswith(".thinking"):
        return "thinking"
    if event.endswith(".error") or event == "graph.error":
        return "error"
    if event == "hitl.interrupted" or event.endswith(".gate"):
        return "gate"
    return "info"


def cap_payload(
    payload: dict[str, Any],
    max_bytes: int = MAX_BYTES,
    max_field: int = MAX_FIELD,
) -> dict[str, Any]:
    """Bound a payload's serialized size.

    Three tiers: return unchanged if it already fits; else truncate any oversize
    string field to a ``{_truncated, _bytes}`` stub; if that still overflows,
    collapse to a single summary stub. Mirrors the Workers port exactly.
    """
    if len(json.dumps(payload).encode()) <= max_bytes:
        return payload

    capped: dict[str, Any] = {}
    for key, value in payload.items():
        if isinstance(value, str) and len(value.encode()) > max_field:
            capped[key] = {"_truncated": True, "_bytes": len(value.encode())}
        else:
            capped[key] = value

    if len(json.dumps(capped).encode()) <= max_bytes:
        return capped

    return {
        "_truncated": True,
        "_bytes": len(json.dumps(payload).encode()),
        "_keys": sorted(payload.keys()),
    }


def derive_duration_ms(
    *,
    event: str,
    step: str | None,
    recorded_at_ms: float,
    last_start_ms: float | None,
) -> int | None:
    """Milliseconds between a ``.done`` event and the most recent matching
    ``.start`` for the same ``(stream_id, step)``.

    Only ``.done`` events carry a duration; everything else (incl. ``.start``
    and ``.thinking``) is ``None``. Returns ``None`` when no matching start was
    seen.
    """
    if not event.endswith(".done"):
        return None
    if last_start_ms is None:
        return None
    return round(recorded_at_ms - last_start_ms)


def persist_thinking_enabled() -> bool:
    """Read the ``PERSIST_THINKING`` toggle (default ON)."""
    raw = os.environ.get("PERSIST_THINKING")
    if raw is None:
        return True
    return raw.strip().lower() in ("1", "true", "on")


def _parse_timestamp_ms(value: object) -> float:
    """Parse an ISO-8601 envelope timestamp into epoch milliseconds.

    Accepts a trailing ``Z`` (RunExecutor emits ``...isoformat() + "Z"``) and
    falls back to ``now`` for an unparseable/missing value so a malformed
    timestamp never blocks persistence.
    """
    if isinstance(value, str) and value:
        normalized = value.replace("Z", "+00:00") if value.endswith("Z") else value
        try:
            return datetime.fromisoformat(normalized).timestamp() * 1000.0
        except ValueError:
            pass
    return datetime.now(UTC).timestamp() * 1000.0


_INSERT_SQL = text(
    """
    INSERT INTO content_tool.run_event_logs
        (log_id, stream_id, stream_kind, seq, event, level, step,
         iteration, duration_ms, payload, recorded_at)
    VALUES
        (:log_id, :stream_id, :stream_kind, :seq, :event, :level, :step,
         :iteration, :duration_ms, CAST(:payload AS jsonb), :recorded_at)
    """
)

_MAX_SEQ_SQL = text(
    "SELECT COALESCE(MAX(seq), -1) FROM content_tool.run_event_logs "
    "WHERE stream_id = CAST(:stream_id AS uuid)"
)


class RunEventLogWriter:
    """Async, batched, fire-and-forget event-log writer.

    Rows are enqueued non-blockingly; a background drain task bulk-inserts when
    the queue reaches ``_FLUSH_THRESHOLD`` or every ``_FLUSH_INTERVAL_S``. Per
    stream the ``seq`` counter is seeded lazily from
    ``COALESCE(MAX(seq), -1) + 1``. A ``.start`` event's recorded time is
    tracked per ``(stream_id, step)`` so the matching ``.done`` can compute
    ``duration_ms``.
    """

    def __init__(self, session_factory: SessionFactory) -> None:
        self._sf = session_factory
        self._queue: asyncio.Queue[dict[str, Any]] = asyncio.Queue()
        self._seq: dict[str, int] = {}
        self._last_start_ms: dict[tuple[str, str], float] = {}
        self._task: asyncio.Task[None] | None = None
        # Serializes flush()/_write_batch so the background drain task and a
        # run's finally-flush cannot race on the shared _seq / _last_start_ms
        # dicts (one writer instance is shared across all concurrent runs).
        # Created lazily so it binds to the running loop, not import-time.
        self._flush_lock: asyncio.Lock | None = None

    def _get_flush_lock(self) -> asyncio.Lock:
        if self._flush_lock is None:
            self._flush_lock = asyncio.Lock()
        return self._flush_lock

    # -- lifecycle ---------------------------------------------------------
    def start(self) -> None:
        """Start the background drain task (idempotent)."""
        if self._task is None or self._task.done():
            self._task = asyncio.create_task(self._drain_loop())

    async def aclose(self) -> None:
        """Flush remaining rows and stop the drain task."""
        await self.flush()
        if self._task is not None and not self._task.done():
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
        self._task = None

    # Alias kept for callers that expect ``close`` (parity with other writers).
    async def close(self) -> None:
        await self.aclose()

    # -- enqueue -----------------------------------------------------------
    def enqueue(self, stream_id: str, stream_kind: str, event_envelope: dict[str, Any]) -> None:
        """Non-blocking enqueue of one SSE envelope for persistence.

        Respects ``PERSIST_THINKING``: a thinking event is dropped here when the
        toggle is off (it is still streamed live by the caller). Any error in
        building/queuing the row is swallowed — persistence must never break the
        emit path.
        """
        try:
            event = str(event_envelope.get("event", ""))
            level = derive_level(event)
            if level == "thinking" and not persist_thinking_enabled():
                return
            self._queue.put_nowait(
                {
                    "stream_id": stream_id,
                    "stream_kind": stream_kind,
                    "event": event,
                    "level": level,
                    "envelope": event_envelope,
                }
            )
        except Exception:
            logger.warning("event_log_enqueue_failed", event=event_envelope.get("event"))

    # -- draining ----------------------------------------------------------
    async def _drain_loop(self) -> None:
        while True:
            await asyncio.sleep(_FLUSH_INTERVAL_S)
            if not self._queue.empty():
                await self.flush()

    async def flush(self) -> None:
        """Drain everything currently queued and bulk-insert it.

        Guarded by ``_flush_lock`` so the background drain and a run's
        finally-flush serialize their access to the shared ``_seq`` /
        ``_last_start_ms`` state. DB errors are logged and the batch dropped —
        never raised into the caller.
        """
        async with self._get_flush_lock():
            batch: list[dict[str, Any]] = []
            while not self._queue.empty():
                try:
                    batch.append(self._queue.get_nowait())
                except asyncio.QueueEmpty:
                    break
            if not batch:
                return
            try:
                await self._write_batch(batch)
            except Exception:
                logger.warning("event_log_flush_failed", count=len(batch))

    async def _seed_seq(self, session: AsyncSession, stream_id: str) -> None:
        if stream_id in self._seq:
            return
        result = await session.execute(_MAX_SEQ_SQL, {"stream_id": stream_id})
        max_seq = result.scalar_one()
        self._seq[stream_id] = int(max_seq) + 1

    async def _write_batch(self, batch: list[dict[str, Any]]) -> None:
        rows: list[dict[str, Any]] = []
        async with self._sf() as session:
            for item in batch:
                stream_id = item["stream_id"]
                await self._seed_seq(session, stream_id)
                rows.append(self._build_row(item))
            await session.execute(_INSERT_SQL, rows)
            await session.commit()

    def _build_row(self, item: dict[str, Any]) -> dict[str, Any]:
        stream_id: str = item["stream_id"]
        event: str = item["event"]
        envelope: dict[str, Any] = item["envelope"]
        step = derive_step(event)
        recorded_at_ms = _parse_timestamp_ms(envelope.get("timestamp"))

        last_start = self._last_start_ms.get((stream_id, step)) if step else None
        duration_ms = derive_duration_ms(
            event=event,
            step=step,
            recorded_at_ms=recorded_at_ms,
            last_start_ms=last_start,
        )
        if event.endswith(".start") and step is not None:
            self._last_start_ms[(stream_id, step)] = recorded_at_ms

        iteration = envelope.get("iteration")
        raw_payload = envelope.get("payload")
        payload: dict[str, Any] = (
            cast("dict[str, Any]", raw_payload) if isinstance(raw_payload, dict) else {}
        )
        seq = self._seq[stream_id]
        self._seq[stream_id] = seq + 1

        return {
            "log_id": str(uuid4()),
            "stream_id": stream_id,
            "stream_kind": item["stream_kind"],
            "seq": seq,
            "event": event,
            "level": item["level"],
            "step": step,
            "iteration": int(iteration) if isinstance(iteration, int) else None,
            "duration_ms": duration_ms,
            "payload": json.dumps(cap_payload(payload), ensure_ascii=False),
            "recorded_at": datetime.fromtimestamp(recorded_at_ms / 1000.0),
        }


# Emitter ContextVar -------------------------------------------------------
# Mirrors ``content_tool.gemini.streaming.set_thought_emitter`` so a node
# wrapper can emit lifecycle events (``*.start`` / ``*.error``) through whatever
# transport the executor bound, without threading the executor through the graph.
EventEmitter = Callable[[str, dict[str, Any]], Awaitable[None]]

_event_emitter: ContextVar[EventEmitter | None] = ContextVar(
    "content_tool_event_emitter", default=None
)


def set_event_emitter(emitter: EventEmitter | None) -> None:
    _event_emitter.set(emitter)


def get_event_emitter() -> EventEmitter | None:
    return _event_emitter.get()


def logged_node[NodeFn: Callable[..., Awaitable[dict[str, Any]]]](
    phase: str,
    name: str,
    node: NodeFn,
) -> NodeFn:
    """Wrap a LangGraph node coroutine to emit lifecycle markers.

    Emits ``f"{phase}.{name}.start"`` before running ``node`` and, on exception,
    ``f"{phase}.{name}.error"`` with ``{error_type, message}`` before re-raising.
    The existing ``*.done`` events (driven by the astream updates loop in the
    executor) are intentionally left intact.

    The wrapper is typed to return the SAME callable type it received so the
    graph builders' ``add_node`` calls typecheck exactly as they did before
    (the wrapper is transparent to pyright).
    """

    async def _wrapped(state: object) -> dict[str, Any]:
        emitter = get_event_emitter()
        if emitter is not None:
            await emitter(f"{phase}.{name}.start", {})
        try:
            return await node(state)
        except Exception as exc:
            if emitter is not None:
                await emitter(
                    f"{phase}.{name}.error",
                    {"error_type": type(exc).__name__, "message": str(exc)},
                )
            raise

    return cast("NodeFn", _wrapped)

File: deploy/cloudflare-workers/src/db/event-log.ts
// ---------------------------------------------------------------------------
// Verbose per-step event log — pure derivation helpers + bulk insert.
//
// Every SSE event that flows through the RUN_STREAM Durable Object is parsed
// into a `content_tool.run_event_logs` row here. These helpers are PURE (no DB,
// no I/O) so they unit-test cleanly and stay byte-identical to the Python
// reference (`content_tool/observability/event_log.py`) and the frontend's
// understanding of the same contract.
//
// The DO owns the seq counter and the per-step ".start" timing map; this module
// only derives row shape from a single envelope + that caller-supplied context.
// ---------------------------------------------------------------------------

import type { getSql } from "./client";

type Sql = ReturnType<typeof getSql>;

/** Event-name verbs whose presence as the LAST segment means the segment
 *  before it is the node/step name (e.g. `strategy.outline.done` → "outline"). */
const VERBS = new Set(["start", "done", "error", "thinking", "completed", "interrupted"]);

const MAX_BYTES = 16_384;
const MAX_FIELD = 2_048;

export type LogLevel = "info" | "thinking" | "gate" | "error";
export type StreamKind = "run" | "batch";

/** The set of levels accepted by the read-API `level` filter. Anything else is
 *  ignored so a garbage query string never reaches the SQL filter. */
const LOG_LEVELS: readonly LogLevel[] = ["info", "thinking", "gate", "error"];

function isLogLevel(value: string): value is LogLevel {
  return (LOG_LEVELS as readonly string[]).includes(value);
}

/** Coerce a postgres.js numeric (string when fetch_types:false) to number|null. */
function toNumberOrNull(value: unknown): number | null {
  if (value === null || value === undefined) {
    return null;
  }
  if (typeof value === "number") {
    return Number.isFinite(value) ? value : null;
  }
  if (typeof value === "string") {
    if (value.trim() === "") {
      return null;
    }
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}

/** The SSE envelope as POSTed to the DO `/append` route. */
export interface EventEnvelope {
  event: string;
  run_id?: string;
  batch_id?: string;
  timestamp: string;
  iteration?: number | null;
  payload?: unknown;
}

/** A derived `run_event_logs` row, ready to bulk-insert. */
export interface EventLogRow {
  stream_id: string;
  stream_kind: StreamKind;
  seq: number;
  event: string;
  level: LogLevel;
  step: string | null;
  iteration: number | null;
  duration_ms: number | null;
  payload: unknown;
  recorded_at: string;
}

/** Mutable map of step → epoch-ms of the most recent ".start" for that step,
 *  threaded by the DO so a later ".done" can compute its duration. */
export type StartMsByStep = Record<string, number>;

// ---------------------------------------------------------------------------
// Pure derivation
// ---------------------------------------------------------------------------

function byteLen(value: string): number {
  return new TextEncoder().encode(value).length;
}

/** Derive the `step` column from an event name. For verb-suffixed events the
 *  step is the segment before the verb; otherwise it is the final segment. */
export function deriveStep(event: string): string | null {
  const segments = event.split(".");
  const last = segments[segments.length - 1];
  if (last !== undefined && VERBS.has(last)) {
    return segments.length > 1 ? (segments[segments.length - 2] ?? null) : null;
  }
  return segments.length > 0 ? (segments[segments.length - 1] ?? null) : null;
}

/** Derive the `level` column from an event name. */
export function deriveLevel(event: string): LogLevel {
  if (event.endsWith(".thinking")) return "thinking";
  if (event.endsWith(".error") || event === "graph.error") return "error";
  if (event === "hitl.interrupted" || event.endsWith(".gate")) return "gate";
  return "info";
}

/** Whether this event should be PERSISTED given the PERSIST_THINKING toggle.
 *  Thinking events are always broadcast; they are only persisted when on. */
export function shouldPersist(event: string, persistThinking: boolean): boolean {
  if (event.endsWith(".thinking")) return persistThinking;
  return true;
}

/** Bound a payload to MAX_BYTES via per-field then whole-payload truncation. */
export function capPayload(payload: unknown): unknown {
  if (payload === null || typeof payload !== "object" || Array.isArray(payload)) {
    // Non-object payloads: only summarize if the encoded form is oversized.
    const serialized = JSON.stringify(payload);
    if (serialized === undefined || byteLen(serialized) <= MAX_BYTES) {
      return payload;
    }
    return { _truncated: true, _bytes: byteLen(serialized) };
  }

  const entries = Object.entries(payload as Record<string, unknown>);
  const full = JSON.stringify(payload);
  if (full !== undefined && byteLen(full) <= MAX_BYTES) {
    return payload;
  }

  // Per-field cap: replace each oversized string field with a stub.
  const capped: Record<string, unknown> = {};
  for (const [key, value] of entries) {
    if (typeof value === "string" && byteLen(value) > MAX_FIELD) {
      capped[key] = { _truncated: true, _bytes: byteLen(value) };
    } else {
      capped[key] = value;
    }
  }
  const cappedSerialized = JSON.stringify(capped);
  if (cappedSerialized !== undefined && byteLen(cappedSerialized) <= MAX_BYTES) {
    return capped;
  }

  // Still too big: whole-payload summary.
  return {
    _truncated: true,
    _bytes: full === undefined ? 0 : byteLen(full),
    _keys: entries.map(([k]) => k).sort(),
  };
}

/**
 * Derive a full `run_event_logs` row from one envelope.
 *
 * `seq` is supplied by the caller (the DO's monotonic counter). `startMsByStep`
 * is mutated in place: a ".start" event records its epoch-ms keyed by step; a
 * ".done" event reads (and clears) the matching start to compute `duration_ms`.
 */
export function parseEventToRow(
  streamId: string,
  streamKind: StreamKind,
  envelope: EventEnvelope,
  seq: number,
  startMsByStep: StartMsByStep,
): EventLogRow {
  const event = envelope.event;
  const step = deriveStep(event);
  const recordedAt = envelope.timestamp;
  const recordedMs = Date.parse(recordedAt);

  let durationMs: number | null = null;
  if (event.endsWith(".start") && step !== null) {
    startMsByStep[step] = recordedMs;
  } else if (event.endsWith(".done") && step !== null) {
    const startMs = startMsByStep[step];
    if (startMs !== undefined && Number.isFinite(recordedMs)) {
      durationMs = recordedMs - startMs;
      delete startMsByStep[step];
    }
  }

  return {
    stream_id: streamId,
    stream_kind: streamKind,
    seq,
    event,
    level: deriveLevel(event),
    step,
    iteration: envelope.iteration ?? null,
    duration_ms: durationMs,
    payload: capPayload(envelope.payload ?? {}),
    recorded_at: recordedAt,
  };
}

// ---------------------------------------------------------------------------
// Persistence (bulk insert, idempotent on (stream_id, seq))
// ---------------------------------------------------------------------------

/**
 * Bulk-insert event rows. `ON CONFLICT (stream_id, seq) DO NOTHING` makes the
 * write idempotent so a Workflow replay (which re-emits memoized events) or a
 * retried DO alarm never double-inserts. JSON payloads are stringified so
 * postgres.js binds them as `jsonb` text.
 */
export async function persistEvents(sql: Sql, rows: readonly EventLogRow[]): Promise<void> {
  if (rows.length === 0) {
    return;
  }
  const values = rows.map((r) => ({
    stream_id: r.stream_id,
    stream_kind: r.stream_kind,
    seq: r.seq,
    event: r.event,
    level: r.level,
    step: r.step,
    iteration: r.iteration,
    duration_ms: r.duration_ms,
    payload: JSON.stringify(r.payload),
    recorded_at: r.recorded_at,
  }));

  await sql`
    INSERT INTO content_tool.run_event_logs ${sql(
      values,
      "stream_id",
      "stream_kind",
      "seq",
      "event",
      "level",
      "step",
      "iteration",
      "duration_ms",
      "payload",
      "recorded_at",
    )}
    ON CONFLICT (stream_id, seq) DO NOTHING
  `;
}

// ---------------------------------------------------------------------------
// Read API
// ---------------------------------------------------------------------------

/** Default + hard cap on the number of rows a single /logs read returns. */
export const DEFAULT_LOG_LIMIT = 2_000;
export const MAX_LOG_LIMIT = 10_000;

export interface LogQuery {
  sinceSeq?: number;
  limit?: number;
  level?: string;
}

/** A row as returned by the read API (keys match the shared contract). */
export interface LogResponseRow {
  log_id: string;
  stream_id: string;
  stream_kind: string;
  seq: number;
  event: string;
  level: string;
  step: string | null;
  iteration: number | null;
  duration_ms: number | null;
  payload: unknown;
  recorded_at: string;
}

/** The raw shape a row arrives in from postgres.js. With `fetch_types:false`,
 *  numeric columns (seq, iteration, duration_ms) come back as strings. */
export interface RawLogRow {
  log_id: string;
  stream_id: string;
  stream_kind: string;
  seq: string | number;
  event: string;
  level: string;
  step: string | null;
  iteration: string | number | null;
  duration_ms: string | number | null;
  payload: unknown;
  recorded_at: string;
}

/**
 * Pure mapper from a raw DB row to the read-API response shape. Coerces the
 * postgres.js numeric strings (`seq`, `iteration`, `duration_ms`) to
 * `number | null` so the frontend renders durations as time, not literal
 * strings like "2500ms". Extracted as a pure function so it unit-tests without
 * a database.
 */
export function mapLogRow(row: RawLogRow): LogResponseRow {
  return {
    log_id: row.log_id,
    stream_id: row.stream_id,
    stream_kind: row.stream_kind,
    seq: toNumberOrNull(row.seq) ?? 0,
    event: row.event,
    level: row.level,
    step: row.step,
    iteration: toNumberOrNull(row.iteration),
    duration_ms: toNumberOrNull(row.duration_ms),
    payload: row.payload,
    recorded_at: row.recorded_at,
  };
}

/** Clamp a requested limit into [1, MAX_LOG_LIMIT], defaulting when unset. */
export function clampLimit(limit: number | undefined): number {
  if (limit === undefined || !Number.isFinite(limit) || limit <= 0) {
    return DEFAULT_LOG_LIMIT;
  }
  return Math.min(Math.trunc(limit), MAX_LOG_LIMIT);
}

/**
 * Read a stream's event log ordered by seq ASC. Supports `since_seq`
 * (seq > sinceSeq), a `limit` (clamped), and a `level` equality filter. `seq`
 * and `duration_ms` are coerced from postgres.js strings to numbers.
 */
export async function getEventLogs(
  sql: Sql,
  streamId: string,
  query: LogQuery,
): Promise<LogResponseRow[]> {
  const limit = clampLimit(query.limit);
  const sinceClause =
    query.sinceSeq !== undefined ? sql`AND seq > ${query.sinceSeq}` : sql``;
  const levelClause = query.level !== undefined ? sql`AND level = ${query.level}` : sql``;

  const rows = await sql<RawLogRow[]>`
    SELECT log_id, stream_id, stream_kind, seq, event, level, step,
           iteration, duration_ms, payload, recorded_at
    FROM content_tool.run_event_logs
    WHERE stream_id = ${streamId}
      ${sinceClause}
      ${levelClause}
    ORDER BY seq ASC
    LIMIT ${limit}
  `;

  return rows.map(mapLogRow);
}

/** Parse the query string of a /logs request into a typed LogQuery. */
export function parseLogQuery(searchParams: URLSearchParams): LogQuery {
  const query: LogQuery = {};
  const since = searchParams.get("since_seq");
  if (since !== null && since.trim() !== "") {
    const parsed = Number(since);
    if (Number.isFinite(parsed)) {
      query.sinceSeq = Math.trunc(parsed);
    }
  }
  const limit = searchParams.get("limit");
  if (limit !== null && limit.trim() !== "") {
    const parsed = Number(limit);
    if (Number.isFinite(parsed)) {
      query.limit = parsed;
    }
  }
  const level = searchParams.get("level");
  if (level !== null) {
    const normalized = level.trim();
    // Only accept a known level; ignore garbage so it never reaches the SQL filter.
    if (isLogLevel(normalized)) {
      query.level = normalized;
    }
  }
  return query;
}
